llpc/util/llpcElfWriter.cpp

```diff
@@ -195,6 +195,57 @@ void ElfWriter<Elf>::mergeMapItem(llvm::msgpack::MapDocNode &destMap, llvm::msgp
   }
 }
 
+// =====================================================================================================================
+// Update descriptor offset to USER_DATA in metaNote, in place in the messagepack document.
+//
+// @param context : context related to ElfNote
+// @param document : [in, out] The parsed message pack document of the metadata note.
+static void updateRootDescriptorRegisters(Context *context, msgpack::Document &document) {
+  auto pipeline = document.getRoot().getMap(true)[Util::Abi::PalCodeObjectMetadataKey::Pipelines].getArray(true)[0];
+  auto registers = pipeline.getMap(true)[Util::Abi::PipelineMetadataKey::Registers].getMap(true);
+  const unsigned mmSpiShaderUserDataVs0 = 0x2C4C;
+  const unsigned mmSpiShaderUserDataPs0 = 0x2c0c;
+  const unsigned mmComputeUserData0 = 0x2E40;
+  unsigned userDataBaseRegisters[] = {mmSpiShaderUserDataVs0, mmSpiShaderUserDataPs0, mmComputeUserData0};
+  unsigned psUserDataCount = context->getGfxIpVersion().major < 9 ? 16 : 32;
+  for (auto baseRegister : userDataBaseRegisters) {
+    for (unsigned i = 0; i < psUserDataCount; ++i) {
```

Comment:
The userDataCount must be per-stage just like the base register: compute still only has 16 user data registers.